File: kotlin/lib/src/main/kotlin/desertroad/solar/Event.kt
package desertroad.solar

/**
 * Events triggering transitions in horizon phases.
 */
enum class Event(val rising: Boolean, val prevPhase: Phase, val nextPhase: Phase) {
    ASTRONOMICAL_DAWN(
        rising = true,
        prevPhase = Phase.NIGHT,
        nextPhase = Phase.ASTRONOMICAL_TWILIGHT
    ),
    NAUTICAL_DAWN(
        rising = true,
        prevPhase = Phase.ASTRONOMICAL_TWILIGHT,
        nextPhase = Phase.NAUTICAL_TWILIGHT
    ),
    CIVIL_DAWN(
        rising = true,
        prevPhase = Phase.NAUTICAL_TWILIGHT,
        nextPhase = Phase.CIVIL_TWILIGHT
    ),
    SUNRISE(
        rising = true,
        prevPhase = Phase.CIVIL_TWILIGHT,
        nextPhase = Phase.DAY
    ),
    SUNSET(
        rising = false,
        prevPhase = Phase.DAY,
        nextPhase = Phase.CIVIL_TWILIGHT
    ),
    CIVIL_DUSK(
        rising = false,
        prevPhase = Phase.CIVIL_TWILIGHT,
        nextPhase = Phase.NAUTICAL_TWILIGHT
    ),
    NAUTICAL_DUSK(
        rising = false,
        prevPhase = Phase.NAUTICAL_TWILIGHT,
        nextPhase = Phase.ASTRONOMICAL_TWILIGHT
    ),
    ASTRONOMICAL_DUSK(
        rising = false,
        prevPhase = Phase.ASTRONOMICAL_TWILIGHT,
        nextPhase = Phase.NIGHT
    );

    val setting get() = !rising
}

File: dart/lib/src/event.dart
import 'package:local_solar/src/phase.dart';

enum Event implements Comparable<Event> {
  astronomicalDawn(
      rising: true,
      prevPhase: Phase.night,
      nextPhase: Phase.astronomicalTwilight),
  nauticalDawn(
      rising: true,
      prevPhase: Phase.astronomicalTwilight,
      nextPhase: Phase.nauticalTwilight),
  civilDawn(
      rising: true,
      prevPhase: Phase.nauticalTwilight,
      nextPhase: Phase.civilTwilight),
  sunrise(rising: true, prevPhase: Phase.civilTwilight, nextPhase: Phase.day),
  sunset(rising: false, prevPhase: Phase.day, nextPhase: Phase.civilTwilight),
  civilDusk(
      rising: false,
      prevPhase: Phase.civilTwilight,
      nextPhase: Phase.nauticalTwilight),
  nauticalDusk(
      rising: false,
      prevPhase: Phase.nauticalTwilight,
      nextPhase: Phase.astronomicalTwilight),
  astronomicalDusk(
      rising: false,
      prevPhase: Phase.astronomicalTwilight,
      nextPhase: Phase.night);

  final bool rising;
  final Phase prevPhase;
  final Phase nextPhase;

  bool get setting => !rising;

  const Event(
      {required this.rising, required this.prevPhase, required this.nextPhase});

  @override
  int compareTo(Event other) => index.compareTo(other.index);
}
